<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Plato - public/constants.js</title>

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="">
  <meta name="author" content="">

  <!--[if lt IE 9]>
  <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
  <![endif]-->

  <link href="../../assets/css/vendor/morris.css" rel="stylesheet">
  <link href="../../assets/css/vendor/bootstrap.css" rel="stylesheet">
  <link href="../../assets/css/vendor/font-awesome.css" rel="stylesheet">
  <link href="../../assets/css/vendor/codemirror.css" rel="stylesheet">
  <link href="../../assets/css/plato.css" rel="stylesheet">
  <link href="../../assets/css/plato-file.css" rel="stylesheet">

</head>

<body>

<div class="navbar navbar-fixed-top">
  <div class="container">
    <a class="navbar-brand" href="http://github.com/es-analysis/plato">Plato on Github</a>
    <ul class="nav navbar-nav">
      <li>
        <a href="../../index.html">Report Home</a>
      </li>
    </ul>
  </div>
</div>

<div class="jumbotron">
  <div class="container">
    <h1>public/constants.js</h1>
  </div>
</div>

<div class="container aggregate-stats">
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Maintainability <a href="http://blogs.msdn.com/b/codeanalysis/archive/2007/11/20/maintainability-index-range-and-meaning.aspx"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="A value between 0 and 100 that represents the relative ease of maintaining the code. A high value means better maintainability." data-original-title="Maintainability Index"  data-container="body"></i></a></h2>
      <p class="stat">19.10</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Lines of code <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h2>
      <p class="stat">326</p>
    </div>
  </div>
  <div class="row historical">
    <div class="col-md-6">
      <p id="chart_historical_maint" class="chart"></p>
    </div>
    <div class="col-md-6">
      <p id="chart_historical_sloc" class="chart"></p>
    </div>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h2 class="header">Difficulty  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="The difficulty measure is related to the difficulty of the program to write or understand." data-original-title="Difficulty" data-container="body"></i></a></h2>
      <p class="stat">8.37</p>
    </div>
    <div class="col-md-6">
      <h2 class="header">Estimated Errors  <a href="http://en.wikipedia.org/wiki/Halstead_complexity_measures"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Halstead's delivered bugs is an estimate for the number of errors in the implementation." data-original-title="Delivered Bugs" data-container="body"></i></a></h2>
      <p class="stat">4.78</p>
    </div>
  </div>
</div>

<div class="container charts">
  <div class="row">
    <h2 class="header">Function weight</h2>
  </div>
  <div class="row">
    <div class="col-md-6">
      <h3 class="chart-header">By Complexity <a href="http://en.wikipedia.org/wiki/Cyclomatic_complexity"><i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="This metric counts the number of distinct paths through a block of code. Lower values are better." data-original-title="Cyclomatic Complexity" data-container="body"></i></a></h3>
      <div id="fn-by-complexity" class="stat"></div>
    </div>
    <div class="col-md-6">
      <h3 class="chart-header">By SLOC  <i class="icon icon-info-sign" rel="popover" data-placement="top" data-trigger="hover" data-content="Source Lines of Code / Logical Lines of Code" data-original-title="SLOC/LSLOC" data-container="body"></i></h3>
      <div id="fn-by-sloc" class="stat"></div>
    </div>
  </div>
</div>

<div class="container">
  <div class="row">
    <textarea id="file-source" class="col-md-12">var path = require(&quot;path&quot;);


var constants = {};
///////////////////////////////////////////
///////////////////////////////////////////
//// DO NOT CHANGE ANYTHING BELOW HERE ////
///////////////////////////////////////////
///////////////////////////////////////////

constants.blast_db      = &#039;ALL_SEQS&#039;;
constants.download_file_formats = [&#039;metadata&#039;,&#039;fasta&#039;,&#039;taxbytax&#039;,&#039;taxbyref&#039;,&#039;taxbyseq&#039;,&#039;biom&#039;,&#039;matrix&#039;,&#039;phyloseq&#039;];
constants.ENV_SOURCE = { 
        10: &quot;air&quot;,
         20: &quot;extreme habitat&quot;,
        30: &quot;host associated&quot;,
         40: &quot;human associated&quot;,
		 41: &quot;human-skin&quot;,
		 42: &quot;human-oral&quot;,
		 43: &quot;human-gut&quot;,
         44: &quot;human-vaginal&quot;,
		 45: &quot;human-amniotic-fluid&quot;,
		 46: &quot;human-urine&quot;,
		 47: &quot;human-blood&quot;,	
	     50: &quot;microbial mat/biofilm&quot;,
		 60: &quot;miscellaneous_natural_or_artificial_environment&quot;,
		 70: &quot;plant associated&quot;,
		 80: &quot;sediment&quot;,
		 90: &quot;soil/sand&quot;,
		 100: &quot;unknown&quot;,
		 110: &quot;wastewater/sludge&quot;,
		 120: &quot;water-freshwater&quot;,
		 130: &quot;water-marine&quot;,
		 140: &quot;indoor&quot;
    };
constants.DOMAINS = { domains: [
        {id: 1, name: &quot;Archaea&quot;},
        {id: 2, name: &quot;Bacteria&quot;},
        {id: 3, name: &quot;Eukarya&quot;},
        {id: 4, name: &quot;Organelle&quot;},
        {id: 5, name: &quot;Unknown&quot;}
    ]};
constants.TARGETS = [&quot;Av3&quot;,&quot;Av6&quot;, &quot;Av6v4&quot;,&quot;Av4v5&quot;, &quot;Av3v5&quot;,&quot;Av5v6&quot;,
                     &quot;Bv1v3&quot;,&quot;Bv3&quot;,&quot;Bv6&quot;, &quot;Bv6v4&quot;,&quot;Bv4v5&quot;, &quot;Bv3v5&quot;,&quot;Bv5v6&quot;,
        &quot;Ev9&quot; ];
constants.UNITSELECT = { units: [
        {id : &#039;tax_silva108_simple&#039;,file: &#039;unit_selectors/taxa_silva108_simple.html&#039;, name : &quot;Taxonomy (Silva-108) -Simple&quot;, subtext: &#039;Silva108&#039;},
        {id : &#039;tax_silva108_custom&#039;,file: &#039;unit_selectors/taxa_silva108_custom.html&#039;, name : &quot;Taxonomy (Silva-108) -Custom&quot;, subtext: &#039;Silva108&#039;},
     //   {id : &#039;tax_silva108_custom_fancytree&#039;,file: &#039;unit_selectors/taxa_silva108_custom_fancytree.json&#039;, name : &quot;Taxonomy -Custom_fancytree&quot;, subtext: &#039;Silva108&#039;},
     //   {id : &#039;tax_silva108_custom_dhtmlx&#039;,file: &#039;unit_selectors/taxa_silva108_custom_dhtmlx.json&#039;, name : &quot;Taxonomy -Custom_dhtmlx&quot;, subtext: &#039;Silva108&#039;},
        {id : &#039;tax_gg_simple&#039;,      file: &#039;unit_selectors/taxa_gg_simple.html&#039;,       name : &quot;TODO-Taxonomy (Greengenes-13.5) -Simple&quot;, subtext: &#039;Greengenes v13.5&#039;},
        {id : &#039;tax_gg_custom&#039;,      file: &#039;unit_selectors/taxa_gg_cust.html&#039;,         name : &quot;TODO-Taxonomy (Greengenes-13.5) -Custom&quot;, subtext: &#039;Greengenes v13.5&#039;},
        {id : &#039;tax_rdp&#039;,            file: &#039;unit_selectors/taxa_rdp.html&#039;,             name : &quot;TODO-Taxonomy RDP&quot;,     subtext: &#039;Release 11&#039;},
        //{id : &#039;TODO-otus&#039;,               file: &#039;unit_selectors/otus.html&#039;,                 name : &quot;OTUs&quot;,             subtext: &#039;ClosedRef GG v13.5&#039;},
        //{id : &#039;TODO-otus&#039;,               file: &#039;unit_selectors/otus.html&#039;,                 name : &quot;OTUs&quot;,             subtext: &#039;UCLUST&#039;},
        {id : &#039;otus&#039;,               file: &#039;unit_selectors/otus.html&#039;,                 name : &quot;TODO-OTUs&quot;,             subtext: &#039;SLP&#039;},
        {id : &#039;med_nodes&#039;,          file: &#039;unit_selectors/med_nodes.html&#039;,            name : &quot;TODO-MED Nodes&quot;,        subtext: &#039;&#039;}
    ]};
constants.UNIT_ASSIGNMENT_CHOICES = { 
        &#039;refRDP_2.12-16S&#039;: { taxonomy_curator:&#039;RDP (2.12) 16S-rRNA&#039;, method:&#039;RDP&#039;,  reference_db:&#039;Default (no training)&#039;, availability:&#039;available&#039;,     refdb:&#039;2.12&#039; },
        &#039;refRDP_2.12-ITS&#039;: { taxonomy_curator:&#039;RDP (2.12) ITS-UNITE&#039;, method:&#039;RDP&#039;,  reference_db:&#039;Default (no training)&#039;, availability:&#039;available&#039;,     refdb:&#039;2.12&#039; },
        &#039;refssu&#039;:   { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;refssu (full-length)&#039;,  availability:&#039;available&#039;,     refdb:&#039;refssu&#039; },
        &#039;refv1v3&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv1v3 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv1v3&#039; },
        &#039;refv3a&#039;:   { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Av3 (Archaeal)&#039;,        availability:&#039;available&#039;, refdb:&#039;refv3a&#039; },
        &#039;refv3&#039;:    { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv3 (Bacterial)&#039;,       availability:&#039;available&#039;, refdb:&#039;refv3&#039; },
        &#039;refv3v5&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv3v5 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv3v5&#039; },
        &#039;refv3v6&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv3v6 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv3v6&#039; },
        &#039;refv4&#039;:    { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv4 (Bacterial)&#039;,       availability:&#039;available&#039;, refdb:&#039;refv4&#039; },
        &#039;refv4v5a&#039;: { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Av4v5 (Archaeal)&#039;,      availability:&#039;available&#039;, refdb:&#039;refv4v5a&#039; },
        &#039;refv4v5&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv4v5 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv4v5&#039; },
        &#039;refv4v6a&#039;: { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Av4v6 (Archaeal)&#039;,      availability:&#039;available&#039;, refdb:&#039;refv4v6a&#039; },
        &#039;refv4v6&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv4v6 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv4v6&#039; },
        &#039;refv5&#039;:    { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv5 (Bacterial)&#039;,       availability:&#039;available&#039;, refdb:&#039;refv5&#039; },
        &#039;refv5v6&#039;:  { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv5v6 (Bacterial)&#039;,     availability:&#039;available&#039;, refdb:&#039;refv5v6&#039; },
        &#039;refv6a&#039;:   { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Av6 (Archaeal)&#039;,        availability:&#039;available&#039;, refdb:&#039;refv6a&#039; },
        &#039;refv6&#039;:    { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Bv6 (Bacterial)&#039;,       availability:&#039;available&#039;, refdb:&#039;refv6&#039; },
        &#039;refv9&#039;:    { taxonomy_curator:&#039;SILVA (v119)&#039;, method:&#039;GAST&#039;, reference_db:&#039;Ev9 (Eukaryal)&#039;,        availability:&#039;available&#039;, refdb:&#039;refv9&#039; },
        &#039;refits1&#039;:  { taxonomy_curator:&#039;UNITE&#039;, method:&#039;GAST&#039;, reference_db:&#039;ITS1&#039;,                         availability:&#039;available&#039;, refdb:&#039;refits1&#039; },
        &#039;refGG_MAY2013&#039;: { taxonomy_curator:&#039;GreenGenes (May2013)&#039;, method:&#039;GAST&#039;, reference_db:&#039;refssu&#039;,        availability:&#039;not available&#039;, refdb:&#039;GG_MAY2013&#039; }
        
        
    };
constants.VISUALOUTPUTCHOICES = { choices: [
        {id : &#039;counts_table&#039;,   show: &#039;Counts Table&#039;},
        {id : &#039;barcharts&#039;,      show: &#039;Counts Bar Charts&#039;},
        {id : &#039;heatmap&#039;,        show: &#039;Distance Heatmap&#039;},
        {id : &#039;dendrogram&#039;,     show: &#039;Community Dendrogram&#039;},
        {id : &#039;alphadiversity&#039;, show: &#039;Alpha Diversity&#039;}
    ]};

constants.NORMALIZATIONCHOICES = { choices: [
        {id: &#039;none&#039;,            show: &#039;Not Normalized (default)&#039;},
        {id: &#039;maximum&#039;,         show: &#039;Normalized to the Maximum Sample&#039;},
        {id: &#039;frequency&#039;,       show: &#039;Normalized to Frequency&#039;}
    ]};

   
constants.DISTANCECHOICES = { choices: [
        
    {id: &#039;jaccard&#039;,         show: &#039;Jaccard&#039;     },
    {id: &#039;kulczynski&#039;,      show: &#039;Kulczynski&#039;  },
    {id: &#039;canberra&#039;,        show: &#039;Canberra&#039;    },
    {id: &#039;morisita_horn&#039;,   show: &#039;Morisita-Horn&#039;},
        // both R and python
        

    {id: &#039;bray_curtis&#039;,     show: &#039;Bray-Curtis&#039; },
 //       {id: &#039;manhattan&#039;,       show: &#039;Manhattan&#039;   },
 //       {id: &#039;gower&#039;,           show: &#039;Gower&#039;       },
 //       {id: &#039;euclidean&#039;,       show: &#039;Euclidean&#039;   },
       
       
 //       {id: &#039;pearson&#039;,         show: &#039;Pearson&#039;     },
 //       {id: &#039;spearman&#039;,        show: &#039;Spearman&#039;    },
// R only
//        {id: &#039;correlation&#039;,     show: &#039;Correlation&#039; },
//        {id: &#039;mountford&#039;,       show: &#039;Mountford&#039;   },
//        {id: &#039;chao_j&#039;,          show: &#039;Chao J&#039;      },
//        {id: &#039;chao_s&#039;,          show: &#039;Chao S&#039;      },
//        {id: &#039;raup&#039;,            show: &#039;Raup&#039;        },
        
//        {id: &#039;yue_clayton&#039;,     show: &#039;Yue-Clayton&#039; }

// python distances:

        // {id: &#039;abund_jaccard&#039;,         show: &#039;Jaccard  - Abundance&#039;     },
        // {id: &#039;binary_jaccard&#039;,        show: &#039;Jaccard - Binary&#039;     },
        // {id: &#039;soergel&#039;,         show: &#039;Soergel&#039;     },
        // {id: &#039;hellinger&#039;,         show: &#039;Hellinger&#039;     },
        // {id: &#039;chord&#039;,         show: &#039;chord&#039;     },
        // {id: &#039;chisq&#039;,         show: &#039;Chisq&#039;     }

    ]};

constants.PORTALS = {
    &#039;CMP&#039;:{ name:&#039;Coral Microbiome Project&#039;, 
            subtext:&#039;CMP&#039;, 
            projects:[&#039;LTR_MCR_Bv6&#039;,&#039;LTR_MCR_Av6&#039;,&#039;LTR_MCR_Ev9&#039;,&#039;ICM_CCB_Bv6&#039;,&#039;ICM_CCB_Av6&#039;],
            prefixes:[&#039;CMP&#039;],
            suffixes:[],
            pagetitle:&#039;Coral Microbe Project Portal&#039;,
            maintitle:&#039;Coral Microbiome Portal&#039;,
            subtitle: &#039;The Coral Microbiome Portal (CMP) database brings together next generation sequencing data of coral-associated microorganisms from studies conducted thoughout the world’s reefs.&#039;,
            thumb:&#039;/images/portals/cmp_thumbnail.jpg&#039;,
            zoom:3  
        },
    &#039;CODL&#039;:{name:&#039;Census of Deep Life&#039;, 
            subtext:&#039;CoDL&#039;, 
            projects:[],
            prefixes:[&#039;DCO&#039;],
            suffixes:[],
            pagetitle:&#039;Census of Deep Life Portal&#039;,
            maintitle:&#039;Census of Deep Life Portal&#039;,
            subtitle: &#039;The mandate of the Census of Deep Life is to perform a global survey of life in continental and marine subsurface environments using deep DNA sequencing technology.&#039;,
            thumb:&#039;/images/portals/dco_thumbnail.jpg&#039;,
            zoom:2  // worldwide
        },
    &#039;HMP&#039;: {name:&#039;Human Mircrobiome Project&#039;, 
            subtext:&#039;HMP&#039;, 
            projects:[],
            prefixes:[&#039;HMP&#039;],
            suffixes:[],
            pagetitle:&#039;Human Microbiome Project Portal&#039;,
            maintitle:&#039;HMP Portal&#039;,
            subtitle: &#039;The NIH Human Microbiome Project is one of several international efforts designed to take advantage of large scale, high through multi ‘omics analyses to study the microbiome in human health.&#039;,
            thumb:&#039;/images/portals/hmp_logo_NIH_retina.png&#039;,
            zoom:4  // mostly US? Do we even have or want distribution?
        },
    &#039;ICOMM&#039;: {name:&#039;International Census of Marine Microbes&#039;, 
            subtext:&#039;ICoMM&#039;, 
            projects:[],
            prefixes:[&#039;ICM&#039;,&#039;KCK&#039;],
            suffixes:[],
            pagetitle: &#039;International Census of Marine Microbes Portal&#039;,
            maintitle: &#039;ICoMM - Microbis Portal&#039;,
            subtitle: &#039;The role of the International Census of Marine Microbes (ICoMM) is to promote an agenda and an environment that will accelerate discovery,&lt;br&gt;understanding, and awareness of the global significance of marine microbes.&#039;,
            thumb:&#039;/images/portals/icomm_thumbnail.jpg&#039;,
            zoom: 2  // worldwide
        },
    &#039;LTER&#039;: {name:&#039;Long Term Ecological Research (LTER)&#039;, 
            subtext:&#039;LTER&#039;, 
            projects:[],
            prefixes:[&#039;LTR&#039;],
            suffixes:[],
            pagetitle: &#039;Microbial Inventory Research Across Diverse Aquatic Sites (MIRADA) Portal&#039;,
            maintitle: &#039;MIRADA Portal&#039;,
            subtitle: &#039;Microbial Inventory Research Across Diverse Aquatic Long Term Ecological Research (LTER) Sites.&#039;,
            thumb:&#039;/images/portals/lter_thumbnail.jpg&#039;,
            zoom: 5  // mostly US
        },
    &#039;MBE&#039;: {name:&#039;Microbiology of the Built Environment&#039;, 
            subtext:&#039;MBE&#039;, 
            projects:[],
            prefixes:[&#039;MBE&#039;],
            suffixes:[],
            pagetitle: &#039;Microbiology Of the Built Environment Portal&#039;,
            maintitle: &#039;MoBEDAC Portal&#039;,
            subtitle: &#039;Microbiome of the Built Environment -Data Analysis Core.&#039;,
            thumb:&#039;/images/portals/mbe_thumbnail.gif&#039;,
            zoom: 4  // mostly US?
        }, 
    &#039;PSPHERE&#039;:{name:&#039;The Plastisphere&#039;, 
            subtext:&#039;Plastisphere&#039;, 
            projects:[&#039;LAZ_SEA_Bv6&#039;,&#039;LAZ_SEA_Ev9&#039;,&#039;LAZ_SEA_Bv6v4&#039;,&#039;LAZ_DET_Bv3v4&#039;],
            prefixes:[],
            suffixes:[],
            pagetitle: &#039;The Plastisphere&#039;,
            maintitle: &#039;Plastisphere Portal&#039;,
            subtitle: &#039;Bacteria and Plastics&#039;,
            thumb:&#039;/images/portals/psphere_thumbnail.jpg&#039;,
            zoom: 5  // mostly US
        },
    &#039;RARE&#039;:{name:&#039;The Rare Biosphere&#039;, 
            subtext:&#039;Rare Biosphere&#039;, 
            projects:[],
            prefixes:[&#039;RARE&#039;],
            suffixes:[],
            pagetitle: &#039;The Rare Biosphere Portal&#039;,
            maintitle: &#039;Rare Biosphere Portal&#039;,
            subtitle: &#039;A New Paradigm for Microbiology.&#039;,
            thumb:&#039;/images/portals/rare_thumbnail.png&#039;,
            zoom: 13  // mostly Falmouth
        },
    &#039;UC&#039;:  {name:&#039;Ulcerative Colitis&#039;, 
            subtext:&#039;Ulcerative Colitis&#039;, 
            projects:[],
            prefixes:[&#039;UC&#039;],
            suffixes:[],
            pagetitle: &#039;Ulcerative Colitis Portal&#039;,
            maintitle: &#039;Ulcerative Colitis Portal&#039;,
            subtitle: &#039;The Role of the Gut Microbiota in Ulcerative Colitis (NIH Human Microbiome Demonstration Project).&#039;,
            thumb:&#039;/images/portals/uc_thumbnail.jpg&#039;,
            zoom: 4  // mostly US?
        },
    &#039;UNIEUK&#039;: {name:&#039;UniEuk&#039;, 
            subtext:&#039;UniEuk&#039;, 
            projects:[],
            prefixes:[],
            suffixes:[&#039;Ev4&#039;,&#039;Ev9&#039;,&#039;Euk&#039;],
            pagetitle: &#039;UniEuk Portal&#039;,
            maintitle: &#039;UniEuk Portal&#039;,
            subtitle: &#039;A Gathering of all Eukaryal Projects.&#039;,
            thumb:&#039;/images/portals/unieuk_thumbnail.jpg&#039;,
            zoom: 2  // worldwide
        }
    }

// This List MUST match the fields in sequence_uniq_infos
constants.AVAILABLE_UNITS = [&#039;silva_taxonomy_info_per_seq_id&#039;, &#039;oligotype_id&#039;, &#039;gg_otu_id&#039;];
// blue to red
constants.HEATMAP_COLORS = [&#039;1111ff&#039;,&#039;3333ff&#039;,&#039;5555ff&#039;,&#039;7777ff&#039;,&#039;9999ff&#039;,&#039;aaaaff&#039;,&#039;ccccff&#039;,&#039;ddeeee&#039;,&#039;eeeedd&#039;,&#039;ffdddd&#039;,&#039;ffbbbb&#039;,&#039;ff9999&#039;,&#039;ff7777&#039;,&#039;ff5555&#039;,&#039;ff3333&#039;,&#039;ff0000&#039;];

constants.RSCRIPT_CMD = &#039;RScript --no-restore --no-save&#039;;

constants.RANKS = [&quot;domain&quot;, &quot;phylum&quot;, &quot;klass&quot;, &quot;order&quot;, &quot;family&quot;, &quot;genus&quot;, &quot;species&quot;, &quot;strain&quot;];

constants.PCT_RANGE = [0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1,2,3,4,5,6,7,8,9,10,20,30,40,50,60,70,80,90,100];

constants.VISUAL_THUMBNAILS = { visuals: [
    {name:&#039;Taxonomy Frequency Table&#039;,   thumb:&#039;/images/visuals/counts_table.png&#039;,   link:&#039;user_viz_data/counts_table&#039;,   id:&#039;counts_table_link_id&#039;, 
        tip:&#039;&#039;  },
    
    {name:&#039;Metadata Table&#039;,             thumb:&#039;/images/visuals/metadata.png&#039;,       link:&#039;user_viz_data/metadata_table&#039;, id:&#039;metadata_table_link_id&#039;, 
        tip:&#039;&#039;},    
    
    {name:&#039;Distance Heatmap (py)&#039;,      thumb:&#039;/images/visuals/heatmap.png&#039;,		link:&#039;user_viz_data/heatmap&#039;,        id:&#039;dheatmap_link_id&#039;, 
        tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)&#039;},
    
    {name:&#039;Stackbar Charts (d3/svg)&#039;,   thumb:&#039;/images/visuals/barcharts.png&#039;,   	link:&#039;user_viz_data/barcharts&#039;,      id:&#039;barcharts_link_id&#039;, 
        tip:&#039;&#039;    },
    
    {name:&#039;Pie Charts (d3/svg)&#039;,        thumb:&#039;/images/visuals/pie_charts.png&#039;,  	link:&#039;user_viz_data/piecharts&#039;,      id:&#039;piecharts_link_id&#039;, 
        tip:&#039;&#039;    },
    
    {name:&#039;Frequency Heatmap (R/pdf)&#039;,  thumb:&#039;/images/visuals/fheatmap.png&#039;,       link:&#039;user_viz_data/frequency_heatmap&#039;, id:&#039;fheatmap_link_id&#039;, 
        tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)-|-R (https://www.r-project.org/)-|-pheatmap (R-package)-|-vegan (R-package)-|-jsonlite (R-package)-|-RColorBrewer (R-package)&#039;},
    
    {name:&#039;Data Browser (Krona)&#039;,       thumb:&#039;/images/visuals/krona.png&#039;,  		link:&#039;user_viz_data/dbrowser&#039;,       id:&#039;dbrowser_link_id&#039;, 
        tip:&#039;&#039;     },
	
    {name:&#039;Dendrogram (d3/phylogram)&#039;,  thumb:&#039;/images/visuals/dendrogram.png&#039;,		link:&#039;user_viz_data/dendrogram&#039;,     id:&#039;dendrogram1_link_id&#039;, 
        tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)&#039;   },
	
    //{name:&#039;Dendrogram (d3-phylonator)&#039;, thumb:&#039;/images/visuals/dendrogram.png&#039;,  	link:&#039;user_viz_data/dendrogram&#039;,     id:&#039;dendrogram2_link_id&#039;, 
   //     tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)&#039;   },
    
    {name:&#039;Dendrogram (d3/radial)&#039;,     thumb:&#039;/images/visuals/radial.png&#039;,         link:&#039;user_viz_data/dendrogram&#039;,     id:&#039;dendrogram3_link_id&#039;, 
        tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)&#039;   },
   

   // {name:&#039;Dendrogram (py-pdf)&#039;,        thumb:&#039;/images/visuals/dendrogram.png&#039;,  	link:&#039;user_viz_data/dendrogram&#039;,     id:&#039;dendrogram_pdf_link_id&#039;, 
   //     tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)-|-matplotlib (python library)&#039;},
    
    {name:&#039;PCoA 2D Analyses (R/pdf)&#039;,   thumb:&#039;/images/visuals/pcoa.png&#039;,        	link:&#039;user_viz_data/pcoa&#039;,           id:&#039;pcoa_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)-|-vegan (R-package)-|-ape (R-package)&#039;        },
    
    {name:&#039;PCoA 3D Analyses (Emperor)&#039;, thumb:&#039;/images/visuals/emperor.png&#039;,        link:&#039;user_viz_data/pcoa&#039;,           id:&#039;pcoa_3d_link_id&#039;, 
        tip:&#039;Python2.7-|-scikit-bio (python library)-|-scipy (python library)-|-numpy (python library)-|-cogent (python library)-|-QIIME (http://qiime.org)&#039; },
    
    {name:&#039;Geo Distribution&#039;,          thumb:&#039;/images/visuals/map.png&#039;,         	link:&#039;user_viz_data/geospatial&#039;,     id:&#039;geospatial_link_id&#039;, 
        tip:&#039;lat/lon metadata&#039;    },
    
    {name:&#039;Alpha Diversity&#039;,            thumb:&#039;/images/visuals/alpha.png&#039;,          link:&#039;user_viz_data/alpha_diversity&#039;,id:&#039;adiversity_link_id&#039;, 
        tip:&#039;Python2.7-|-scipy (python library)-|-numpy (python library)-|-scikit-bio&#039;},
    
    {name:&#039;Phyloseq Bars (R/svg)&#039;,      thumb:&#039;/images/visuals/phyloseq_bars.png&#039;,  link:&#039;user_viz_data/phyloseq01&#039;,  id:&#039;phyloseq01_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)&#039;},
    
    {name:&#039;Phyloseq Heatmap (R/png)&#039;,   thumb:&#039;/images/visuals/phyloseq_heatmap.png&#039;,link:&#039;user_viz_data/phyloseq02&#039;, id:&#039;phyloseq02_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)&#039;},
    
    {name:&#039;Phyloseq Network (R/svg)&#039;,   thumb:&#039;/images/visuals/phyloseq_network.png&#039;,link:&#039;user_viz_data/phyloseq03&#039;, id:&#039;phyloseq03_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)&#039;},
    
    {name:&#039;Phyloseq Ordination (R/svg)&#039;,thumb:&#039;/images/visuals/phyloseq_ord1.png&#039;,  link:&#039;user_viz_data/phyloseq04&#039;, id:&#039;phyloseq04_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)&#039;},
    
    {name:&#039;Phyloseq Tree (R/svg)&#039;,      thumb:&#039;/images/visuals/phyloseq_tree.png&#039;,     link:&#039;user_viz_data/phyloseq05&#039;, id:&#039;phyloseq05_link_id&#039;, 
        tip:&#039;R (https://www.r-project.org/)-|-phyloseq (R-package)&#039;}
   
]};
    
constants.REQ_METADATA_FIELDS = [&quot;altitude&quot;,  &quot;collection_date&quot;, &quot;common_name&quot;, &quot;country&quot;, &quot;depth&quot;, &quot;description&quot;, &quot;elevation&quot;, &quot;env_biome&quot;, &quot;env_feature&quot;, &quot;env_matter&quot;, &quot;latitude&quot;, &quot;longitude&quot;, &quot;public&quot;, &quot;taxon_id&quot;];
constants.CONTACT_US_SUBJECTS = [&quot;Account Request&quot;, &quot;Report a Problem&quot;, &quot;Announce a Publication&quot;, &quot;Other&quot;];

module.exports = constants;</textarea>
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p>.</p>
  </div>
</footer>

<script type="text/html" id="complexity-popover-template">
  <div class="complexity-notice">
    Complexity : {{ complexity.cyclomatic }} <br>
    Length : {{ complexity.halstead.length }} <br>
    Difficulty : {{ complexity.halstead.difficulty.toFixed(2) }} <br>
    Est # bugs : {{ complexity.halstead.bugs.toFixed(2) }}<br>
  </div>
</script>

<script type="text/javascript" src="../../assets/scripts/bundles/core-bundle.js"></script>
<script type="text/javascript" src="../../assets/scripts/bundles/codemirror.js"></script>
<script type="text/javascript" src="../../assets/scripts/codemirror.markpopovertext.js"></script>
<script type="text/javascript" src="report.js"></script>
<script type="text/javascript" src="report.history.js"></script>
<script type="text/javascript" src="../../assets/scripts/plato-file.js"></script>
</body>
</html>
